docs/Boxes.md

```diff
@@ -0,0 +1,143 @@
+
+## The `Boxes` module {#sec-boxes}
+
+A _Box_ datatype is one whose values can hold either a success indication, optionally along with a value, or a failure indication, optionally with error information.
+These are particularly useful with Dafny's abrupt-termination-on-failure `:-` operator.
+
+Any user datatype can serve this purpose, as long as it has an `IsFailure?` predicate 
+(in which case it is known as a [_failure-compatible_](https://dafny.org/latest/DafnyRef/DafyRef#TODO) type --- an FC-type). 
+In this module Dafny defines three such types, illustrates them with examples, and then describes 
+how they can be used in a system where different parts of the system use different FC-types.
+
+The three types are
+- `Option<R>` - which is either `Some` with a value of type `T` or a `None` with no information
+- `Outcome<E>` - which is either `Pass` with no informatino or `Fail` with an error value of type `E` (often a `string`) 
+- `Result<R,E>` - which is either `Success` with a value of type `R` or `Failure` with error value of type `E` (often a `string`)
```

Comment:
I will also note that `Outcome` is primarily useful with `Need`.
Without this function it is of dubious use.
And I don't see any documentation on `Need` here,
which I would think of as a miss since it is primarily useful
when dealing with runtime checks for `Result` types.

Reply:
Included text and an example of Need